### PR TITLE
Add definitions about copy commands with textures

### PR DESCRIPTION
**Pull request by Jiawei-Shao**
This patch adds several definitions that are required in the
validation rules of copy commands with textures.
- the internal slots of a GPUTexture
- texel block
- texel block size
- texel block width
- texel block height
- the physical size of a texture subresource
- GPUTextureCopyView

This patch also adds a sub-title for GPUImageBitmapCopyView to
keep it in the same style as GPUBufferCopyView.

***

### :boom: Error: 500 Internal Server Error :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Mar 23, 2020, 9:25 PM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [HTML Diff Service](http://services.w3.org/htmldiff) - The HTML Diff Service is used to create HTML diffs of the spec changes suggested in a pull request.

:link: [Related URL](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2Fgpuweb%2Fgpuweb%2Fpull%2F623%2F2ce1971.html&doc2=https%3A%2F%2Fpr-preview.s3.amazonaws.com%2FJiawei-Shao%2Fgpuweb%2Fpull%2F623.html)

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at 
 sysreq@w3.org to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
</body></html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20gpuweb/gpuweb%23623.)._
</details>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -410,6 +410,15 @@ The [=subresources=] of textures included in the views provided to
 are considered to have {{GPUTextureUsage/OUTPUT_ATTACHMENT}}
 for the [=usage scope=] of this render pass.
 
+The <dfn dfn>physical size</dfn> of a {{GPUTexture}} [=subresource=] is the dimension of the {{GPUTexture}}
+[=subresource=] in pixels that includes the possible extra paddings to form complete [=texel blocks=] in the
+[=subresource=]. The extra paddings only exists in a compressed texture mipmap whose {{GPUTexture/[[textureSize]]}}
+at [=mipmap level=] 0 are multiple of [=texel block size=], but its subdivided levels are not. For example,
```
Comment (Kangz) [inline spec/index.bs]: > The extra paddings only exists in a compressed texture mipmap whose {{GPUTexture/[[textureSize]]}}
at [=mipmap level=] 0 are multiple of [=texel block size=], but its subdivided levels are not.

This seems to be a conditional statement, that says that padding exists only if level 0 is a multiple of the texel block size. How about a more explicit explanation like:

Compressed textures always have a mipmap level 0 whose extent is a multiple of the texel block size, but lower mipmap levels might not be multiple of the texel block size and can have padding.

File: spec/index.bs [hunk 2]
```diff
@@ -410,6 +410,15 @@ The [=subresources=] of textures included in the views provided to
...
+considering a texture in BC format whose {{GPUTexture/[[textureSize]]}} is {60, 60, 1}, when sampling the texture
+at [=mipmap level=] 2, the sampling hardware uses {15, 15, 1} as the size of the [=subresource=], while its
+[=physical size=] is {16, 16, 1} as the block-compression algorithm can only operate on 4x4 [=texel blocks=]. There
+is no extra padding for the {{GPUTexture}} [=subresources=] with non-compressed texture formats.
```
Comment (Kangz) [inline spec/index.bs]: I think it would be more clear to make the difference between block-based texture formats and pixel-based formats because eventually we might have other formats like YUV 420 that are block-based but not traditionally thought of as compressed.

File: spec/index.bs [hunk 3]
```diff
@@ -1322,6 +1331,34 @@ interface GPUTexture {
 GPUTexture includes GPUObjectBase;
 </script>
 
+{{GPUTexture}} has the following internal slots:
+<dl dfn-type=attribute dfn-for="GPUTexture">
+    : <dfn>\[[textureSize]]</dfn> of type {{GPUExtent3D}}.
+    ::
+        The size of the {{GPUTexture}} in texels.
```
Comment (Kangz) [inline spec/index.bs]: The size of mipmap level 0

File: spec/index.bs [hunk 4]
```diff
@@ -1442,6 +1479,16 @@ shader. Compressed texture formats are provided by extensions. Their naming
 should follow the convention here, with the texture name as a prefix. e.g.
 `etc2-rgba8unorm`.
 
+The <dfn dfn>texel block</dfn> is a single addressable element of an uncompressed texture, or a single
```
Comment (Kangz) [inline spec/index.bs]: Same comment about talking about block-based, maybe? @kainino0x @kvark @JusSn what do you think?

File: spec/index.bs [hunk 5]
```diff
@@ -410,6 +410,15 @@ The [=subresources=] of textures included in the views provided to
 are considered to have {{GPUTextureUsage/OUTPUT_ATTACHMENT}}
 for the [=usage scope=] of this render pass.
 
+The <dfn dfn>physical size</dfn> of a {{GPUTexture}} [=subresource=] is the dimension of the {{GPUTexture}}
+[=subresource=] in pixels that includes the possible extra paddings to form complete [=texel blocks=] in the
```
Comment (kvark) [inline spec/index.bs]: "pixels" -> "texels"?

File: spec/index.bs (hunk 1 above)
Comment (kvark) [inline spec/index.bs]: Instead of "For examples," can we use bikesheds formatting like:
```
Example: xxx
```

File: spec/index.bs [hunk 6]
```diff
@@ -1322,6 +1331,34 @@ interface GPUTexture {
...
+    ::
+        The size of the {{GPUTexture}} in texels.
+
+    : <dfn>\[[mipLevelCount]]</dfn> of type {{GPUIntegerCoordinate}}.
+    ::
+        The total number of the mipmap levels of the {{GPUTexture}}.
+
+    : <dfn>\[[sampleCount]]</dfn> of type {{GPUSize32}}.
+    ::
+        The sample count of the {{GPUTexture}}.
```
Comment (kvark) [inline spec/index.bs]: "Number of samples in each texel of the {{GPUTexture}}"

File: spec/index.bs (hunk 4 above)
Comment (kvark) [reply spec/index.bs]: yeah, if we can just have a separation between block based and texel based, only bringing in the compressed formats in examples, that seems to be cleaner

File: spec/index.bs [hunk 7]
```diff
@@ -1442,6 +1479,16 @@ shader. Compressed texture formats are provided by extensions. Their naming
 should follow the convention here, with the texture name as a prefix. e.g.
 `etc2-rgba8unorm`.
 
+The <dfn dfn>texel block</dfn> is a single addressable element of an uncompressed texture, or a single
+compressed block of a compressed texture. The <dfn dfn>texel block size</dfn> of a {{GPUTextureFormat}} is
+the number of bytes to store one [=texel block=]. The [=texel block size=] of each {{GPUTextureFormat}} is a
+constant value except for {{GPUTextureFormat/"depth24plus"}} and {{GPUTextureFormat/"depth24plus-stencil8"}}.
+
+The <dfn dfn>texel block width</dfn> and <dfn dfn> texel block height </dfn> specifies the dimension of
```
Comment (kvark) [inline spec/index.bs]: `height </dfn>` - extra space?

File: spec/index.bs [hunk 8]
```diff
@@ -1442,6 +1479,16 @@ shader. Compressed texture formats are provided by extensions. Their naming
...
+one [=texel block=]. [=texel block width=] is the number of pixels in each row of one [=texel block=],
```
Comment (kvark) [inline spec/index.bs]: number of "texels"

File: spec/index.bs [hunk 9]
```diff
@@ -1442,6 +1479,16 @@ shader. Compressed texture formats are provided by extensions. Their naming
...
+and [=texel block height=] is the number of texel rows in one [=texel block=]. The [=texel block width=]
+and [=texel block height=] of non-compressed texture formats are 1.
```
Comment (kvark) [inline spec/index.bs]: as @Kangz noted, this isn't true for YUV

File: spec/index.bs [hunk 10]
```diff
@@ -2379,8 +2428,16 @@ dictionary GPUTextureCopyView {
 };
 </script>
 
+A {{GPUTextureCopyView}} is a view of a sub-region of a [=texture=] [=subresource=] with the initial offset {{GPUOrigin3D} in
+texels, used when copying data between two [=texture=]s, from a [=buffer=] to a [=texture=] and from a [=texture=] to a [=buffer=].
```
Comment (kvark) [inline spec/index.bs]: could just be "used when copying data to or from a [=texture=]"

File: spec/index.bs [hunk 11]
```diff
@@ -2379,8 +2428,16 @@ dictionary GPUTextureCopyView {
...
+
+  - For {{GPUTextureDimension/2d}} textures, data is copied between one [=image=] and one [=array layer=].
```
Comment (kvark) [inline spec/index.bs]: this is unclear. One "image" *is* an array layer here

File: spec/index.bs (hunk 1 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 2 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 3 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 4 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 5 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 1 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 6 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 7 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 8 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 9 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 10 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 11 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: I've made some changes here. PTAL, thanks!

File: spec/index.bs [hunk 12]
```diff
@@ -410,6 +410,23 @@ The [=subresources=] of textures included in the views provided to
 are considered to have {{GPUTextureUsage/OUTPUT_ATTACHMENT}}
 for the [=usage scope=] of this render pass.
 
+The <dfn dfn>physical size</dfn> of a {{GPUTexture}} [=subresource=] is the dimension of the {{GPUTexture}}
+[=subresource=] in texels that includes the possible extra paddings to form complete [=texel blocks=] in the
+[=subresource=].
+
+  - For pixel-based {{GPUTextureFormat}}s, the [=physical size=] is always equal to the size of the [=subresource=]
+    used in the sampling hardwares.
+  - {{GPUTexture}}s in compressed {{GPUTextureFormat}}s always have a [=mipmap level=] 0 whose {{GPUTexture/[[textureSize]]}}
```
Comment (Kangz) [inline spec/index.bs]: nit: block-based formats here too.

File: spec/index.bs [hunk 13]
```diff
@@ -410,6 +410,23 @@ The [=subresources=] of textures included in the views provided to
...
+    is a multiple of the [=texel block size=], but the lower mipmap levels might not be the multiple of the [=texel block size=]
+    and can have paddings.
+
+<div class="example">
+Considering a {{GPUTexture}} in BC format whose {{GPUTexture/[[textureSize]]}} is {60, 60, 1}, when sampling
+the {{GPUTexture}} at [=mipmap level=] 2, the sampling hardware uses {15, 15, 1} as the size of the [=subresource=],
+while its [=physical size=] is {16, 16, 1} as the block-compression algorithm can only operate on 4x4 [=texel blocks=].
+There is no extra padding for the {{GPUTexture}} [=subresources=] with non-compressed {{GPUTextureFormat}}s.
```
Comment (Kangz) [inline spec/index.bs]: This sentence is redundant with the pixel-based format list item above. I think it could be removed.

File: spec/index.bs [hunk 14]
```diff
@@ -410,6 +410,23 @@ The [=subresources=] of textures included in the views provided to
 are considered to have {{GPUTextureUsage/OUTPUT_ATTACHMENT}}
 for the [=usage scope=] of this render pass.
 
+The <dfn dfn>physical size</dfn> of a {{GPUTexture}} [=subresource=] is the dimension of the {{GPUTexture}}
```
Comment (Kangz) [inline spec/index.bs]: In a way I think we should define what a texel block is, with a definition that explains that regular format have a texel block size of 1. Then we can define pixel-based formats (1x1 blocks), and block-based formats (other block sizes). The discussion of compressed formats would only be in examples.

Then a lot of the discussion can only talk about block sizes in general, and have notes for what happens for pixel-based formats (i.e. no extra padding).

Basically the spec would be written as if all formats are block-based, but would explain what happens in the usual case when the block size is 1x1.

File: spec/index.bs [hunk 15]
```diff
@@ -2379,8 +2438,16 @@ dictionary GPUTextureCopyView {
 };
 </script>
 
+A {{GPUTextureCopyView}} is a view of a sub-region of a [=texture=] [=subresource=] with the initial offset {{GPUOrigin3D}} in
+texels, used when copying data from or to a {{GPUTexture}}.
+
+  - For {{GPUTextureDimension/2d}} textures, data is copied from or to one [=mipmap level=] and one [=array layer=] of the [=texture=].
+  - For {{GPUTextureDimension/3d}} textures, data is copied from or to one [=mipmap level=] of the [=texture=].
```
Comment (kvark) [inline spec/index.bs]: doesn't say anything about a slice of a 3D texture?

File: spec/index.bs (hunk 14 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: The definition of "texel block" is in the Chapter "Texture Formats", which I think is a better place for it. What do you think?

File: spec/index.bs (hunk 13 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 12 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 15 above)
Comment (Jiawei-Shao) [reply spec/index.bs]: Oh I haven't implemented copies with 1D or 3D textures in Dawn so I am not sure about the related validation rules. I've left a TODO here. What do you think?

**Comment by Jiawei-Shao** [comment]
The error message of the build failure is not related to this patch ("bikeshed requires Python '>=3.7' but the running Python is 2.7.15"). PTAL, thanks!

File: spec/index.bs [hunk 16]
```diff
@@ -1442,6 +1486,18 @@ shader. Compressed texture formats are provided by extensions. Their naming
...
 
+The <dfn dfn>texel block</dfn> is a single addressable element of the textures in pixel-based {{GPUTextureFormat}}s,
+and a single compressed block of the textures in block-based compressed {{GPUTextureFormat}}s.
+
+The <dfn dfn>texel block width</dfn> and <dfn dfn>texel block height</dfn> specifies the dimension of one [=texel block=].
+  - For pixel-based {{GPUTextureFormat}}s, the [=texel block width=] and [=texel block height=] are always 1.
+  - For block-based compressed {{GPUTextureFormat}}s, the [=texel block width=] is the number of texels in each row of one [=texel block=],
+    and the [=texel block height=] is the number of texel rows in one [=texel block=].
+
+The <dfn dfn>texel block size</dfn> of a {{GPUTextureFormat}} is the number of bytes to store one [=texel block=].
```
Comment (Kangz) [inline spec/index.bs]: nit: texel block byte size, otherwise it sounds like (width, height)

File: spec/index.bs [hunk 17]
```diff
@@ -410,6 +410,22 @@ The [=subresources=] of textures included in the views provided to
...
 
+The <dfn dfn>physical size</dfn> of a {{GPUTexture}} [=subresource=] is the dimension of the {{GPUTexture}}
+[=subresource=] in texels that includes the possible extra paddings to form complete [=texel blocks=] in the
+[=subresource=].
+
+  - For pixel-based {{GPUTextureFormat}}s, the [=physical size=] is always equal to the size of the [=subresource=]
+    used in the sampling hardwares.
+  - {{GPUTexture}}s in block-based compressed {{GPUTextureFormat}}s always have a [=mipmap level=] 0 whose {{GPUTexture/[[textureSize]]}}
+    is a multiple of the [=texel block size=], but the lower mipmap levels might not be the multiple of the [=texel block size=] and can
+    have paddings.
```
Comment (JusSn) [inline spec/index.bs]:
```suggestion
    have padding.
```

File: spec/index.bs [hunk 18]
```diff
@@ -1442,6 +1486,18 @@ shader. Compressed texture formats are provided by extensions. Their naming
 should follow the convention here, with the texture name as a prefix. e.g.
 `etc2-rgba8unorm`.
 
+The <dfn dfn>texel block</dfn> is a single addressable element of the textures in pixel-based {{GPUTextureFormat}}s,
+and a single compressed block of the textures in block-based compressed {{GPUTextureFormat}}s.
```
Comment (kvark) [inline spec/index.bs]: "compressed block of textures"?

File: spec/index.bs [hunk 19]
```diff
@@ -1442,6 +1486,18 @@ shader. Compressed texture formats are provided by extensions. Their naming
 should follow the convention here, with the texture name as a prefix. e.g.
 `etc2-rgba8unorm`.
 
+The <dfn dfn>texel block</dfn> is a single addressable element of the textures in pixel-based {{GPUTextureFormat}}s,
```
Comment (kvark) [inline spec/index.bs]: "texel", "element", "pixel" - are all referring to the same thing...

Comment (JusSn) [inline spec/index.bs]:
```suggestion
The <dfn dfn>texel block</dfn> is a single addressable element of textures with pixel-based {{GPUTextureFormat}}s,
```

File: spec/index.bs (hunk 18 above)
Comment (JusSn) [inline spec/index.bs]:
```suggestion
and a single compressed block of textures with block-based compressed {{GPUTextureFormat}}s.
```

File: spec/index.bs (hunk 15 above)
Comment (kainino0x) [reply spec/index.bs]: Always OK with me to leave things TODO.

File: spec/index.bs (hunk 14 above)
Comment (kainino0x) [reply spec/index.bs]: It's probably okay if the definition of texel block is in Texture Formats but we reference it from here. The important comment here, though, is about rewording everything so that there's no "pixel-based" vs "block-based" format distinction. It allows us to write a lot of spec text without mentioning the difference at all, which simplifies it and makes it more readable.

**Comment by kainino0x** [comment]
editors meeting resolution: we'll merge this and do revisions in follow-up.

**Comment by jdashg** [comment]
Please rebase after #633

**Comment by kainino0x** [comment]
Pushed a merge so we can get this landed.
